### ISSUE TITLE
Overwriting NaN values in land cells to push particles back to the ocean

### ISSUE THREAD

**Issue by DEHewitt** [OceanParcels/parcels#1018]
Hi,

I am currently working with a ROMS hydrodynamic model to track particles and am having some issues with particles getting "beached". The simulations are for the east coast of Australia and the domain of the grid is roughly lat: -15 to -37.5 and lon: 142 to 160. A fair portion of the cells making up this grid are land cells and consequently have no velocities associated with them (`NaN`'s which become `0`). Our simulations also include diffusion. Below is  the code we're using to set up the field and diffusion:
```
# Set up the hydrodynamic model
data_path = '/srv/scratch/z5278054/portunid_particle_tracking/ozroms/'
ufiles = sorted(glob('/srv/scratch/z5278054/portunid_particle_tracking/ozroms/2*'))
vfiles = ufiles
wfiles = ufiles
tfiles = ufiles
mesh_mask = data_path + 'bathymetry.nc'

filenames = {'U': {'lon': mesh_mask, 'lat': mesh_mask, 'depth': mesh_mask, 'data': ufiles},
             'V': {'lon': mesh_mask, 'lat': mesh_mask, 'depth': mesh_mask, 'data': vfiles},
             'WA': {'lon': mesh_mask, 'lat': mesh_mask, 'depth': mesh_mask, 'data': wfiles},
             'temp': {'lon': mesh_mask, 'lat': mesh_mask, 'depth': mesh_mask, 'data': tfiles},
             'bathy': {'lon': mesh_mask, 'lat': mesh_mask, 'depth': mesh_mask, 'data': mesh_mask}}

variables = {'U': 'u',
             'V': 'v',
             'WA': 'w', 
             'temp': 'sst',
             'bathy': 'h'}

dimensions = {'U': {'lon': 'lon', 'lat': 'lat', 'depth': 'depth', 'time': 'time'},
              'V': {'lon': 'lon', 'lat': 'lat', 'depth': 'depth', 'time': 'time'},
              'WA': {'lon': 'lon', 'lat': 'lat', 'depth': 'depth', 'time': 'time'},
              'temp': {'lon': 'lon', 'lat': 'lat', 'depth': 'depth', 'time': 'time'},
              'bathy': {'lon': 'lon', 'lat': 'lat', 'time': 'time'}}

# Define fieldset
fieldset = FieldSet.from_nemo(filenames, variables, dimensions, allow_time_extrapolation = True)

# Define diffusion constant
Kh_zonal = 8.8 # following Cetina Heredia et al. (2015, 2019)
Kh_meridional = Kh_zonal

# Add constant to field
size2D = (fieldset.U.grid.ydim, fieldset.U.grid.xdim) 
fieldset.add_field(Field('Kh_zonal', Kh_zonal*np.ones(size2D), 
                         lon=fieldset.U.grid.lon, lat=fieldset.U.grid.lat, mesh='spherical'))
fieldset.add_field(Field('Kh_meridional', Kh_meridional*np.ones(size2D), # size3D?
                         lon=fieldset.U.grid.lon, lat=fieldset.U.grid.lat, mesh='spherical'))
```
Note: vertical velocities are named `WA` following #765 to allow vertical advection independent of sigma levels.

Some of our particles end up in the land cells and then jitter around based on the `Kh_zonal` and `Kh_meridional` values, but don't ever move very far and don't get advected due to the `NaN`'s for `U` and `V`.

My question is: is there some way I can overwrite the `NaN` values in `U` of land cells and give them an arbitray velocity oriented in the eastward direction to push any particles back into the ocean?

Thanks for any help you can provide!

**Comment by reint-fischer** [comment]
Hi @DEHewitt ,

The way this is usually implemented is to create an additional field in preprocessing that has velocities away from the coast. This field is generally created by taking a mask of all the land or 'coastal' cells, in your case where the `NaNs` are, and populating that field with values in the direction of the nearest 'ocean' cell. This example shows how:
1. [the field is created](https://github.com/OceanParcels/Mediterranean_inverse_modelling/blob/master/01_run_parcels.py#L520-L563)
2. [the field is loaded into the fieldset](https://github.com/OceanParcels/Mediterranean_inverse_modelling/blob/master/01_run_parcels.py#L941-L994)
3. [a kernel interacts with the field](https://github.com/OceanParcels/Mediterranean_inverse_modelling/blob/master/01_run_parcels.py#L785-L802)

Another example can be found [here](https://github.com/OceanParcels/Parcelsv2.0PaperNorthSeaScripts/blob/master/preprocess/unbeach_Cfield.py)

I hope these help you implement your solution!

**Comment by hkhirsh** [comment]
Hi @DEHewitt and @reint-fischer! I'm trying to do something very similar. I have two velocity fields. I want to use one field for advection most of the time then use the second if I am unbeaching (it is a field where I have estimated an alongshore current to push particle adjacent to land instead of onto it). I'm having a hard time using the new velocity field for advection. The UnBeaching kernel in the third example above looks like what I want to do, but how did you create fieldset.UV_unbeach instead of fieldset.UV? I tried writing something similar to lines 987 and 988 in the second example above, but I get an error that "Field" is not defined. 

```
fname='guaRoms_000.nc'
lname = 'guaROMs_longshore_2021-04-13_10:06.nc'

filenames = {'U':fname, 
             'V':fname, 
             'uAlongShore':lname, 
             'vAlongShore':lname}

#variables = {'U':'u',                #this is how I defined my variables before, but I wasn't sure how to use the alongshore velocity
#             'V':'v', 
#             'uAlongShore':'u', 
#             'vAlongShore':'v'}

variables = {'U':'u', 
             'V':'v'}

dimensions = {'lat': 'latitude', 'lon': 'longitude', 'depth': 'depth', 'time': 'time'}

U_land = Field('U_land',uAlongShore,lon=lons,lat=lats,fieldtype='u')
V_land = Field('V_land',vAlongShore,lon=lons,lat=lats,fieldtype='v')

fieldset = FieldSet.from_netcdf(filenames, variables, dimensions)
fieldset.add_field(U_land)
fieldset.add_field(V_land)
```

**Comment by hkhirsh** [comment]
@DEHewitt @reint-fischer
Oh and I meant to describe my (ideal) strategy for unbeaching.

Kernel 1: uses AdvectionRK4 to advect all particles
Kernel 2: uses DiffusionUniformKh to diffuse all particles that have advected
Kernel 3: test for beaching (essentially the same as BeachTesting in the second example above)
Kernel 4: "Bounce" the particle. I want to bounce the particle back offshore to it's previous position. I'm haven't figured out how to do this yet, but there must be a way to call on the index for the previous timestep and set particle.lat and particle.lon equal to the previous coordinates.
Kernel 5: if the particle has "bounced" then I now advect it using the alongshore velocity field instead.

Basically, if the particle is close enough to land to beach (aka we detect it beaching in Kernel 3), then we can return it to where it was and use the AlongShore field to advect it for that timestep instead.

I'm new to both Parcels and Python so I'm all ears for better ways to do this, and/or ways to actually implement what I am envisioning here! Thank you!

**Comment by DEHewitt** [comment]
Thanks @reint-fischer, as it turns out the number of particles getting beached via advection was pretty low and most of them were actually just being released in 'land' cells due to the way we are selecting points for release. We've currently implemented an approach like #1008 to nudge the genuinely beached particles (i.e. beached via advection) eastward by adding a random increment to the `particle.lon` (luckily we're working with a coastline that runs pretty much north-south), but will likely move to just deleting these. Thanks again for your help!
